sound/soc/sof/pm.c

```diff
@@ -16,8 +16,146 @@
 #include <asm/pgtable.h>
 #include <sound/core.h>
 #include <sound/soc.h>
+#include <sound/sof.h>
+#include "ops.h"
 #include "sof-priv.h"
 
+static int sof_restore_pipelines(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_widget *swidget = NULL;
+	struct snd_sof_route *sroute = NULL;
+	struct sof_ipc_comp_host *host;
+	struct sof_ipc_buffer *buffer;
+	struct sof_ipc_comp_volume *vol;
+	struct sof_ipc_comp_mixer *mixer;
+	struct sof_ipc_pipe_new *pipeline;
+	struct sof_ipc_comp_src *src;
+	struct sof_ipc_comp_tone *tone;
+	struct snd_sof_dai *dai;
+	struct sof_ipc_comp_dai *comp_dai;
+	int ret = 0;
+
+	/* restore pipeline components */
+	list_for_each_entry_reverse(swidget, &sdev->widget_list, list) {
+		struct sof_ipc_comp_reply r;
+
+		switch (swidget->id) {
+		case snd_soc_dapm_dai_in:
+		case snd_soc_dapm_dai_out:
+			dai = (struct snd_sof_dai *)swidget->private;
+			comp_dai = &dai->comp_dai;
+			ret = sof_ipc_tx_message(sdev->ipc,
+						 comp_dai->comp.hdr.cmd,
+						 comp_dai, sizeof(*comp_dai),
+						 &r, sizeof(r));
+			break;
+		case snd_soc_dapm_mixer:
+			mixer = (struct sof_ipc_comp_mixer *)swidget->private;
+			ret = sof_ipc_tx_message(sdev->ipc,
+						 mixer->comp.hdr.cmd, mixer,
+						 sizeof(*mixer), &r,
+						 sizeof(r));
+			break;
+		case snd_soc_dapm_pga:
+			vol = (struct sof_ipc_comp_volume *)swidget->private;
+			ret = sof_ipc_tx_message(sdev->ipc, vol->comp.hdr.cmd,
+						 vol, sizeof(*vol), &r,
+						 sizeof(r));
+			break;
+		case snd_soc_dapm_buffer:
+			buffer = (struct sof_ipc_buffer *)swidget->private;
+			ret = sof_ipc_tx_message(sdev->ipc,
+						 buffer->comp.hdr.cmd, buffer,
+						 sizeof(*buffer), &r,
+						 sizeof(r));
+			break;
+		case snd_soc_dapm_aif_out:
+		case snd_soc_dapm_aif_in:
```

Comment:
need explicit /* fallthrough */ to make tools happy that this is not a programming error